ArduCopter/mode_ship_ops.cpp

```diff
@@ -0,0 +1,754 @@
+    _approach_mode = approach_mode;
+    switch (_approach_mode) {
+    case ApproachMode::LAUNCH_RECOVERY:
+        gcs().send_text(MAV_SEVERITY_WARNING, "ShipOps: Mode Launch/Recovery");
+        break;
+    case ApproachMode::PAYLOAD_PLACE:
+        gcs().send_text(MAV_SEVERITY_WARNING, "ShipOps: Mode Payload Place");
+        break;
+    }
+
+}
+
+void ModeShipOperation::run()
+{
+    float yaw_cd = attitude_control->get_att_target_euler_cd().z;
+    float yaw_rate_cds = 0.0f;
+
+    Vector2f perch_offset = Vector2f(ship_perch_radius * 100.0, 0.0f);
+    perch_offset.rotate(radians(ship_perch_angle));
+    float perch_height = ship_perch_altitude * 100.0;
+
+    // get pilot desired climb rate if enabled
+    bool valid_pilot_input = !copter.failsafe.radio && copter.ap.rc_receiver_present && g.land_repositioning;
+    if (valid_pilot_input) {
+        // get pilot desired climb rate
+        target_climb_rate = get_pilot_desired_climb_rate(channel_throttle->get_control_in());
+        target_climb_rate = constrain_float(target_climb_rate, -get_pilot_speed_dn(), g.pilot_speed_up);
+    } else {
+        // take off and landing will be conducted automaticaly  
+        if (ship_takeoff) {
+            // take off and move to perch
+            target_climb_rate = g.pilot_speed_up;
+        } else {
+            // return to perch and continue to landing
+            target_climb_rate = -get_pilot_speed_dn();
+        }
+    }
+
+    Vector3f pos_delta_ned_m;  // vector to lead vehicle + offset
+    Vector3f pos_delta_with_ofs_ned_m;  // vector to lead vehicle + offset
+    Vector3f pos_with_ofs_ned;  // vector to lead vehicle + offset
+    Vector3f vel_ned_ms;  // velocity of lead vehicle
+    Vector3f accel_ned;  // accel of lead vehicle
+    
+    const Vector3f &curr_pos_neu_cm = inertial_nav.get_position_neu_cm();
+    int32_t alt_home_above_origin_cm;
+
+    if (!AP::ahrs().get_home().get_alt_cm(Location::AltFrame::ABOVE_ORIGIN, alt_home_above_origin_cm)) {
+        alt_home_above_origin_cm = 0;
+    }
+    
+    // define target location
+    if (g2.follow.get_target_dist_and_vel_ned(pos_delta_ned_m, pos_delta_with_ofs_ned_m, vel_ned_ms)) {
+        // vel_ned_ms does not include the change in heading + offset radius
+        pos_with_ofs_ned.xy() = curr_pos_neu_cm.xy() + pos_delta_with_ofs_ned_m.xy() * 100.0;
+        pos_with_ofs_ned.z = -curr_pos_neu_cm.z + pos_delta_with_ofs_ned_m.z * 100.0;
+        vel_ned_ms *= 100.0f;
+
+        float target_heading_deg = 0.0f;
+        g2.follow.get_target_heading_deg(target_heading_deg);
+
+        if (!ship_available) {
+            // reset ship pos, vel, accel to current value when detected.
+            ship_pos_ned = pos_with_ofs_ned.topostype();
+            ship_vel_ned = vel_ned_ms;
+            ship_accel_ned.zero();
+            ship_heading = radians(target_heading_deg);
+            ship_heading_rate = 0.0;
+            ship_heading_accel = 0.0;
+            ship_available = true;
```

Comment:
```suggestion
           ship.reset(pos_with_ofs_ned, vel_ned_ms, target_heading_deg);
```

... and use that same function in `::init()` to initialise the ship model.

Reply:
Might need help here.